src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -676,6 +704,23 @@ void BaseCouplingScheme::doImplicitStep()
     // no convergence achieved for the coupling iteration within the current time window
     if (_acceleration) {
       _acceleration->performAcceleration(getAccelerationData());
+      /**
+       * Acceleration changes CouplingData::values(), so we must override the data in CouplingData::_timeStepsStorage.
+       * There are generally two possibilities:
+       *
+       * 1) Only override the data in CouplingData::_timeStepsStorage that is part of the receive data
+       * 2) Override the data in CouplingData::_timeStepsStorage for all data in getAccelerationData()
+       *
+       * We are using strategy 2), because it's easier to access getAccelerationData() from here. However, this also means that
+       * we have to make sure that the send data is stored in the CouplingData::_timeStepsStorage - even though this is not
+       * needed at the moment. Important note: In https://github.com/precice/precice/pull/1414 also send data requires to keep
+       * track of _timeStepsStorage for subcycling. So it will become simpler as soon as subcycling is fully implemented.
+       */
+      // @todo For other Acceleration schemes as described in "Rüth, B, Uekermann, B, Mehl, M, Birken, P, Monge, A, Bungartz, H-J. Quasi-Newton waveform iteration for partitioned surface-coupled multiphysics applications. Int J Numer Methods Eng. 2021; 122: 5236– 5257. https://doi.org/10.1002/nme.6443" we need a more elaborate implementation.
+      for (auto &pair : getAccelerationData()) {
+        bool mustOverride = true;
+        pair.second->storeValuesAtTime(time::Storage::WINDOW_END, pair.second->values(), mustOverride);
+      }
```

Comment:
Fixed override vs. overwrite in 43c54159f6d0757c2aaefdf2c33058a0138eb102.

In `receiveData` we clear the storage first and then store the values. There is no need for overwriting and all data (except acceleration data) is final. Only for the acceleration data we have to override all data after acceleration, because the acceleration puts the accelerated data into `pair.second->values()`, not into `time::Storage`.

We could also let the acceleration directly access the storage, but this would lead to many changes (again) and probably a lot of spaghetti. (Same argumentation like for the mapping)

Reply:
I changed this part in aa2765b47c6ef81039ca41647f30814884e5ee34. This is strictly not needed, but I think it's easier to read, if we first put data from the storage into `pair.second->values()` before acceleration, then acceleration processed the data and after acceleration the coupling scheme takes care of putting the data at the right place into `pair.second->values()`.

Reply:
Did you take into consideration that acceleration data is used to compute the acceleration (primary data), but then **all** data is accelerated (secondary data)? See also: https://precice.org/configuration-acceleration.html#coupling-data-and-primary-data

Reply:
I guess this might cause trouble as soon as we apply acceleration to a scenario where subcycling is used. Here, I think no special treatment is needed: The acceleration only "knows" about the data returned from in `getAccelerationData()` and we also take care of all this data. If there is a possible problem, we should first come up with an integration test to reproduce this problem. Otherwise we will break it again. I see two possibilities:

* We are missing a test. Then we should add a test on develop, not here.
* We are not missing a test. Then everything should work.

For subcycling i.e. #1414 I'm quite sure we are missing a test here. But let's worry about this as soon as #1504 is merged (already complicated enough).

Reply:
My bad: acceleration data already contains primary and secondary data. So "taking care of acceleration data" should be enough.

I could not find a suitable existing integration or unit test right away. Could be very well that we are lacking tests here.

Reply:
> Could be very well that we are lacking tests here.

Wouldn't be too surprised. I will create make this a todo in #1414, because we definitely want to check it there. Everything will become more complicated with subcycling and without proper testing at least I will get 100% lost.